### PR TITLE
Add optional subcommands

### PR DESCRIPTION
**Pull request by nlsun**
Also introduces a function in the ParseContext that allows the
current action determine if it is the currently executing command.

### CODE REVIEW

**Comment by nlsun** [comment]
@alecthomas Please let me know what you think.

File: cmd.go [hunk 1]
```diff
@@ -236,6 +237,12 @@ func (c *CmdClause) Command(name, help string) *CmdClause {
 	return cmd
 }
 
+// SubcommandOptional makes subcommands optional
+func (c *CmdClause) SubcommandOptional() *CmdClause {
```
Comment (alecthomas) [inline cmd.go]: Could you name this `OptionalSubcommands()`? It reads better in the fluent style.

**Comment by alecthomas** [comment]
Is the idea here that you could have:

```
command subcommand0
command subcommand1
command subcommand2
```

And command execution could terminate at just `command`?

If so, seems reasonable.

Couple of other comments:

- Could you apply this to the v3-unstable branch please? I'm not accepting new features on master (in fact, I think I'll switch them soon).
- This needs tests.

**Comment by nlsun** [comment]
> Is the idea here that you could have:
> 
> ```
> command subcommand0
> command subcommand1
> command subcommand2
> ```
> And command execution could terminate at just command?

yup

> Could you apply this to the v3-unstable branch please? I'm not accepting new features on master (in fact, I think I'll switch them soon).

Ok, I'll open another PR, should this also go in v2?

> This needs tests.
Will do, just wanted to pitch the idea first

**Comment by alecthomas** [comment]
> Ok, I'll open another PR, should this also go in v2?

No, don't worry about it. v2 will be bugfix only soon.

> Will do, just wanted to pitch the idea first

Understood, thanks, I appreciate that.

**Comment by nlsun** [comment]
Superceded by https://github.com/alecthomas/kingpin/pull/196

**Pull request closed without merging**